src/mlpack/core/data/save_impl.hpp

```diff
@@ -197,6 +197,194 @@ bool Save(const std::string& filename,
   return true;
 }
 
+// Save a Sparse Matrix
+template<typename eT>
+bool Save(const std::string& filename,
+          const arma::SpMat<eT>& matrix,
+          const bool fatal,
+          bool transpose)
+{
+  Timer::Start("saving_data");
+
+  // First we will try to discriminate by file extension.
+  std::string extension = Extension(filename);
+  if (extension == "")
+  {
+    Timer::Stop("saving_data");
+    if (fatal)
+      Log::Fatal << "No extension given with filename '" << filename << "'; "
+          << "type unknown.  Save failed." << std::endl;
+    else
+      Log::Warn << "No extension given with filename '" << filename << "'; "
+          << "type unknown.  Save failed." << std::endl;
+
+    return false;
+  }
+
+  // Catch errors opening the file.
+  std::fstream stream;
+#ifdef  _WIN32 // Always open in binary mode on Windows.
+  stream.open(filename.c_str(), std::fstream::out | std::fstream::binary);
+#else
+  stream.open(filename.c_str(), std::fstream::out);
+#endif
+  if (!stream.is_open())
+  {
+    Timer::Stop("saving_data");
+    if (fatal)
+      Log::Fatal << "Cannot open file '" << filename << "' for writing. "
+          << "Save failed." << std::endl;
+    else
+      Log::Warn << "Cannot open file '" << filename << "' for writing; save "
+          << "failed." << std::endl;
+
+    return false;
+  }
+
+  bool unknownType = false;
+  arma::file_type saveType;
+  std::string stringType;
+
+  if (extension == "csv")
+  {
+    saveType = arma::coord_ascii;
+    stringType = "CSV data";
+  }
+  else if (extension == "txt")
+  {
+    saveType = arma::coord_ascii;
+    stringType = "raw ASCII formatted data";
+  }
+  else if (extension == "bin")
+  {
+    saveType = arma::arma_binary;
+    stringType = "Armadillo binary formatted data";
+  }
+  else if (extension == "pgm")
+  {
+    saveType = arma::pgm_binary;
+    stringType = "PGM data";
+  }
+  else if (extension == "h5" || extension == "hdf5" || extension == "hdf" ||
+           extension == "he5")
```

Comment:
I'm not sure why you've included these formats?  Armadillo sparse matrices can't be saved in this format as far as I know.